model/task/task.go

```diff
@@ -376,20 +390,71 @@ func (t *Task) DependenciesMet(depCaches map[string]Task) (bool, error) {
 	if len(depIdsToQueryFor) > 0 {
 		newDeps, err := Find(ByIds(depIdsToQueryFor).WithFields(StatusKey, DependsOnKey))
 		if err != nil {
-			return false, err
+			return nil, errors.WithStack(err)
 		}
 
 		// add queried dependencies to the cache
 		for _, newDep := range newDeps {
 			deps = append(deps, newDep)
-			depCaches[newDep.Id] = newDep
+			depCache[newDep.Id] = newDep
 		}
 	}
 
-	for _, depTask := range deps {
-		if !t.satisfiesDependency(&depTask) {
+	return deps, nil
+}
+
+// DependenciesSatisfiable returns true when a dependency is satisfied
+// or could be satisfied in the future, and false if any of the first
+// order dependencies are blocked, checking the cache if necessary.
+//
+// Use dependencies met (which can also check the cache,) to figure
+// out if a task is runable.
+func (t *Task) DependencySatisfiable(depCache map[string]Task) (bool, error) {
+	if len(t.DependsOn) == 0 || t.OverrideDependencies {
+		return true, nil
+	}
+
+	// do this early, if possible to avoid caching tasks that we
+	// won't need.
+	for _, dep := range t.DependsOn {
+		if dep.Unattainable {
+			return false, nil
+		}
+	}
+
+	deps, err := t.populateDependencyTaskCache(depCache)
+	if err != nil {
+		return false, errors.WithStack(err)
+	}
+
+	for _, dep := range t.DependsOn {
+		depTask, ok := depCache[dep.TaskId]
+		if !ok {
+			grip.Error(message.Fields{
+				"task_id":     t.Id,
+				"dep_task_id": depTask.Id,
+				"cache_size":  len(deps),
+				"operation":   "DependencySatisfiable",
+				"message":     "cache failure",
+			})
+			continue
+		}
+
+		if depTask.Blocked() {
```

Comment:
this too: if Unattainable is false, none of the dependencies should be blocked.

Reply:
What is the case this is checking for over and above what's reflected by Unattainable?

Reply:
I agree with you in principal, but the premise of this is that sometimes a write doesn't go through, or the data is a bit inconsistent, and the new dispatcher (that has dependencies included in the queue) will end up having tasks which will never be attainable, but will always hang out in the queue. this isn't a problem, exactly, because we'll never dispatch them; _however_, it does mean that we'll tend to over-provision hosts that we can't use as a side effect.